mongoimport/common.go

```diff
@@ -414,13 +570,36 @@ func validateFields(fields []string) error {
 				return fmt.Errorf("fields cannot be identical: '%v' and '%v'", field, latterField)
 			}
 		}
+
+		if useArrayIndexFields {
```

Comment:
I also wonder if this should move into the loop above after 571 (possibly factored out into a function), so that each field only checks fields after it.

Reply:
Yes I think we could move it in. My original implementation had to check the whole array which is why it is where it is. 

The other method I considered was building a tree, where each node had a type associated with it. For each node that's added, you can check the types match for all ancestors.

Reply:
Also it looks like this validation is slightly broken. At the top level of a document "0" and "a" should be compatible but this throws an error. So I need to fix that.